app/controllers/interest_controller.rb

```diff
@@ -25,100 +25,133 @@ class InterestController < ApplicationController
+  end
+
+  private
+
+  def find_relevant_notifications
+    Notification.for_user(@user).sort do |a, b|
       result = a.flavor.to_s <=> b.flavor.to_s
       result = a.summary.to_s <=> b.summary.to_s if result.zero?
       result
     end
-    # interests = Interest.find_all_by_user_id(@user.id).sort do |a,b|
-    interests = Interest.where(user_id: @user.id).sort do |a, b|
+  end
+
+  def find_relevant_interests
+    Interest.for_user(@user).sort do |a, b|
       result = a.target_type <=> b.target_type
       if result.zero?
         result = (a.target ? a.target.text_name : "") <=>
                  (b.target ? b.target.text_name : "")
       end
       result
     end
-    @targets = notifications + interests
-    @pages = paginate_numbers(:page, 50)
-    @pages.num_total = @targets.length
-    @targets = @targets[@pages.from..@pages.to]
   end
 
+  public
+
   # Callback to change interest state in an object.
   # Linked from: show_<object> and emails
   # Redirects back (falls back on show_<object>)
   # Inputs: params[:type], params[:id], params[:state], params[:user]
   # Outputs: none
-  def set_interest
+  def set_interest # rubocop:disable Metrics/AbcSize
     pass_query_params
-    type   = params[:type].to_s
-    oid    = params[:id].to_i
-    state  = params[:state].to_i
-    uid    = params[:user]
-    target = Comment.find_object(type, oid)
-    if @user
-      interest = Interest.find_by(
-        target_type: type, target_id: oid, user_id: @user.id
+    target_type = params[:type].to_s
+    target_id   = params[:id].to_i
+    state       = params[:state].to_i
+    user_id     = params[:user]
+    target      = Comment.find_object(target_type, target_id)
+    return unless @user
+
+    interest = Interest.find_by(
+      target_type: target_type, target_id: target_id, user_id: @user.id
+    )
+    if user_id && @user.id != user_id.to_i
+      flash_error(:set_interest_user_mismatch.l)
+    elsif !target && state != 0
+      flash_error(:set_interest_bad_object.l(type: target_type, id: target_id))
+    else
+      set_interest_state_for_target(interest, state, target)
+    end
+    redirect_to_target_or_interests(target, target_id)
+  end
+
+  private
+
+  def redirect_to_target_or_interests(target, target_id)
+    unless target
+      return redirect_back_or_default(controller: "/interest",
+                                      action: "list_interests")
+    end
+
+    redirect_back_or_default(
+      add_query_param(controller: target.show_controller,
+                      action: target.show_action, id: target_id)
+    )
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def set_interest_state_for_target(interest, state, target)
+    if !interest && state != 0
+      interest = Interest.new
+      interest.target = target
+      interest.user = @user
+    end
+    if state.zero?
+      remove_interest_from_target_and_flash_notice(interest, target)
+    elsif interest.state == true && state.positive?
+      flash_notice(
```

Comment:
Got it. Committing a change now.

Question: is there any reason not to make ivars for `@interest`, `@state`, and `@target` here?
Instead of passing them around everywhere?

Reply:
Since you ask... (!)

One purpose of ivars is as a screwy mechanism for "passing" parameters to the view templates.  But that's not needed here.

The other is to store state information in a class.  I mean, yes, you could spin your suggested use as "state information", but it's stretching it a bit.  It smells more like laziness to me.  (No offense intended!  I've done it myself, no judgement!)

I personally think that a better solution in this case is to re-imagine the design so that parameters don't need to be passed around so much.  Whenever lots of parameters are flying all over, it means concerns aren't be separated properly, or we need a class to hold that information (e.g. a "bounding box" class that can hold north, south, east, west extents in a single object instead of forcing you to pass around four parameters in mapping routines).

In this case, note that target already provides access to target.type and target.id, so there go two parameters, nicely!  And login_required assures that @user.id is the same as params[:user_id], so there it goes.  I think you end up with: target, interest and state, right?  And doesn't interest give you access to the target?

Among the other things I like to see in "good" code --in addition to things like well-named methods, well-separated concerns, few parameters, globals all explicitly set (as opposed to being set by side-effect in submethods) -- is to have top-level parameters clearly exposed.  In the case of controllers, your parameters are all in params[:xxx], of course.  And there's no real reason to pass "params" around, as it's already freely available to everyone.  My own code has been all over the board with respect to this.  Basically, if it's convenient and I'm not feeling lazy, I'll pass params[:state] explicitly to the submethods that need it, if it thereby moves all accesses to params[:xxx] to the top level action method.  It's just nice to be able to see exactly what a method's inputs and outputs are.

I'm also fine with simply documenting the list of parameters in a comment.  But note that if you're trying to track down a bug associated with a parameter, or if you're trying to see how an action actually uses a parameter, it can be a pain in the butt drilling down multiple levels to find where the parameter is actually used.  So, I think it is still "best" to move them all into the top-level and pass them around as parameters.  But not critical.

One classic counter-case is whitelisted_parameters.  But we all know to look for parameters there.  That's a standard mechanism for bundling and processing large numbers of parameters all associated with a single model.